File: src/main/kotlin/Day2.kt
class Day2: Day(2) {
    override val title = "Password Philosophy"

    override fun part1(): Int {
        val regex = "(\\d*)-(\\d*) (\\w): (.*)".toRegex()
        return inputLines.count {
            val match = regex.find(it)!!
            val (start, end, letter, password) = match.groups.toList().drop(1).map { it?.value }
            val count = password?.count { letter?.contains(it) ?: false }!!
            start!!.toInt() <= count && count <= end!!.toInt()
        }
    }

    override fun part2(): Int {
        return inputLines.count {
            val regex = "(\\d*)-(\\d*) (\\w): (.*)".toRegex()
            val match = regex.find(it)!!
            val (first, second, letter, password) = match.groups.toList().drop(1).map { it!!.value }

            try {
                (password.elementAt(first.toInt() - 1) == letter.elementAt(0)) xor
                        (password.elementAt(second.toInt() - 1) == letter.elementAt(0))
            } catch (e: StringIndexOutOfBoundsException) {
                false
            }
        }
    }

}

File: python/2020/day2.py
from typing import List
import re
from operator import xor

# https://regex101.com/r/RTevsV/1/ 
PASSWORD_REGEX = r'(\d+)\-(\d+) (\w)\: (\w+)'

def part1(lines: List[str]) -> int:
  valid = 0
  for line in lines:
    match = re.match(PASSWORD_REGEX, line, re.M|re.I) 
    minimum, maximum, letter, password = match.groups()
    minimum = int(minimum)
    maximum = int(maximum)

    count = password.count(letter) 
    if minimum <= count <= maximum:
      valid += 1
  return valid


def part2(lines: List[str]) -> int:
  valid = 0
  for line in lines:
    match = re.match(PASSWORD_REGEX, line, re.M|re.I) 
    first, second, letter, password = match.groups()
    if xor(password[int(first) - 1] == letter, password[int(second) - 1] == letter):
      valid += 1
  return valid
    
    
if __name__ == '__main__':
    with open("../../exercises/2020/02.txt") as file:
        lines = file.readlines()
        print(part1(lines))
        print(part2(lines))
